solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -368,10 +368,152 @@ private StatefulCollectionWatch compute(
+      if (!collectionWatches.isWatched(coll)) {
+        return;
+      }
+      if (log.isTraceEnabled()) {
+        log.trace("an event happened for {}, event: {}", coll, event.toString());
+      }
+      DocCollection collectionState = getCollection(coll);
+      if (collectionPath.equals(event.getPath())) {
+        DocCollection newState = null;
+        try {
+          newState = fetchCollectionState(coll, null);
+        } catch (KeeperException.SessionExpiredException
+            | KeeperException.ConnectionLossException e) {
+          log.warn("ZooKeeper watch triggered, but Solr cannot talk to ZK: ", e);
+        } catch (KeeperException e) {
+          log.error("exception for collection: [{}]", coll, e);
+          throw new ZooKeeperException(ErrorCode.SERVER_ERROR, "A ZK error has occurred", e);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          log.error("exception for collection: [{}]", coll, e);
+        }
+        collectionWatches.updateDocCollection(coll, newState);
+        synchronized (getUpdateLock()) {
+          constructState(Collections.singleton(coll));
+        }
+      }
+      if (collectionState != null && collectionState.isPerReplicaState()) {
+        String path = event.getPath();
+        if ((event.getType() == EventType.NodeCreated || event.getType() == EventType.NodeDeleted)
+            && path.length() > collectionState.getZNode().length()
+            && path.startsWith(collectionPath)) {
+          if (log.isDebugEnabled()) {
+            log.debug("PRS node event : {}", event.getType());
+          }
+
+          PerReplicaStates prs = collectionState.getPerReplicaStates();
+          if (prs != null) {
+            String stateStr = path.substring(collectionPath.length() + 1);
+            PerReplicaStates.State newState = PerReplicaStates.State.parse(stateStr);
+            PerReplicaStates.State oldState = prs.states.get(newState.replica);
+            if (event.getType() == EventType.NodeCreated) {
+              if (oldState != null && newState.version < oldState.version) {
+                // we got a notification out of order? . Shouldn't happen
+                if (log.isTraceEnabled()) {
+                  log.trace("newState {} < oldState {}", newState, oldState);
+                }
+                return;
+              }
+              if (oldState == null) {
+                // the state does not exist now. fetch everything
+
+                if (log.isTraceEnabled()) {
+                  log.trace("fresh replica, force fetch all {}", collectionPath);
+                }
+
+                prs = PerReplicaStatesFetcher.fetch(collectionPath, zkClient, null);
+              } else {
+                // the PRS entry is already available, need to do an update
+                Stat stat = null;
+                try {
+                  // get the stat of this child node
+                  stat = zkClient.exists(path, null, true);
```

Comment:
What if we get stat of `state.json` here?

Reply:
that's incorrect . The stat of `state.json` would've gotten changed by another event